File: src/widget/net-worth-snapshot.ts
import type { Currency } from '../currency/currency';
import { formatMoney } from '../currency/format';
import type { AccountRow, HoldingRow } from '../db/schema';
import { activeLocale } from '../i18n/active-locale';
import { activeHoldings } from '../rates/active-holdings';
import type { RateTable } from '../rates/conversion';
import { sumByCurrency } from '../rates/currency-totals';
import { guardedNetWorth } from '../rates/net-worth-view';

// The snapshot the app writes to the App Group container for the widget to read.
// Its numbers come from the SAME net-worth math the home screen uses (never a
// second computation) so the widget can never disagree with the app.
//
// SECURITY: this file is cleartext JSON in a shared container. It carries the
// MINIMUM the widget actually renders — the formatted total and the per-currency
// breakdown — and nothing else. A 30-day trend series used to be written here
// for a chart the widget no longer draws; it was removed because a wealth
// history on disk widened the exposure for no reader. Do not add a field the
// SwiftUI view does not render. See docs/security/README.md.
export type NetWorthSnapshot = {
  baseCurrency: Currency;
  total: { formatted: string; minorUnits: number };
  breakdown: { currency: Currency; minorUnits: number; formatted: string }[];
  updatedAt: number;
};

export const buildNetWorthSnapshot = (input: {
  holdings: readonly HoldingRow[];
  accounts: readonly Pick<AccountRow, 'id' | 'archivedAt'>[];
  rateTable: RateTable;
  baseCurrency: Currency;
  now: number;
}): NetWorthSnapshot => {
  const active = activeHoldings(input.holdings, input.accounts);
  const total = guardedNetWorth(active, input.baseCurrency, input.rateTable, input.now);
  const breakdown = sumByCurrency(active, input.now).map((money) => ({
    currency: money.currency,
    minorUnits: money.minorUnits,
    formatted: formatMoney(money, activeLocale()),
  }));

  return {
    baseCurrency: input.baseCurrency,
    total: { formatted: formatMoney(total, activeLocale()), minorUnits: total.minorUnits },
    breakdown,
    updatedAt: input.now,
  };
};

File: ios/KikoWidget/NetWorthSnapshot.swift
import Foundation

/// The single shared App Group identifier for the Kiko app and the
/// KikoWidget extension. Must match the entitlement on BOTH targets and the
/// `<APP_GROUP_ID>` used by the native `WidgetBridge` module that writes the
/// snapshot file.
enum WidgetConstants {
    static let appGroupID = "group.com.dmytro.pff"
    static let snapshotFileName = "net-worth-snapshot.json"
}

/// Mirrors `NetWorthSnapshot` in `src/widget/net-worth-snapshot.ts` EXACTLY.
/// The widget never re-derives net worth math; it only decodes and renders
/// the numbers the app already computed.
struct NetWorthSnapshot: Codable {
    struct Total: Codable {
        let formatted: String
        let minorUnits: Int
    }

    struct BreakdownItem: Codable, Identifiable {
        let currency: String
        let minorUnits: Int
        let formatted: String

        // Not part of the wire format; lets SwiftUI iterate without index math.
        var id: String { currency }
    }

    struct TrendPoint: Codable {
        let time: Int
        // `amount` on the TS side (`NetWorthPoint`) is the total in base MAJOR
        // units (a float), not minor units, so this is a Double, not an Int.
        let value: Double
    }

    /// Every user-facing string the widget renders, resolved by the APP through
    /// `i18n.t` at snapshot-build time. Deliberately not a Localizable.strings
    /// bundle: a .lproj bundle follows the DEVICE language, while Kiko's
    /// language is a persisted in-app setting, so the two would disagree.
    /// One entry per string the widget renders FROM a snapshot, no more — the
    /// no-snapshot placeholder below cannot read this, so it holds no key here.
    /// Mirrors `NetWorthSnapshot['labels']` in src/widget/net-worth-snapshot.ts.
    struct Labels: Codable {
        let title: String
    }

    let baseCurrency: String
    let total: Total
    let breakdown: [BreakdownItem]
    let trend: [TrendPoint]
    let updatedAt: Int
    // Non-optional ON PURPOSE, like every other field here: a snapshot written
    // before this shipped has no `labels` key, so it fails to decode,
    // `SnapshotLoader.load()` returns nil, and the widget shows its placeholder
    // until the app writes again (which it does on every backgrounding). That
    // self-healing miss is preferable to an optional that would let a
    // half-localized widget render indefinitely. Do not "fix" it to `Labels?`.
    let labels: Labels
}

enum SnapshotLoader {
    /// Reads and decodes the snapshot the app's `WidgetBridge` wrote into the
    /// shared App Group container. Returns `nil` on ANY failure (missing
    /// container, missing file, unreadable data, malformed JSON) so callers
    /// can fall back to the placeholder — this must never throw into the
    /// timeline provider.
    static func load() -> NetWorthSnapshot? {
        guard
            let containerURL = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: WidgetConstants.appGroupID
            )
        else {
            return nil
        }

        let fileURL = containerURL.appendingPathComponent(WidgetConstants.snapshotFileName)

        guard let data = try? Data(contentsOf: fileURL) else {
            return nil
        }

        return try? JSONDecoder().decode(NetWorthSnapshot.self, from: data)
    }
}
